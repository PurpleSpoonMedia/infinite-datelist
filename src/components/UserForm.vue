<template>
  <div class="px-lg-7">
    <form @submit.prevent="handleSaveUser">
      <div class="form-group" :class="{error: $v.user.name.$error}">
        <label for="name">Name</label>
        <input
          type="text"
          class="form-control"
          id="name"
          aria-describedby="name"
          placeholder="Enter your name"
          v-model.trim="$v.user.name.$model"
        >
        <p v-if="$v.user.name.$error && !$v.user.name.required">{{ errorMessages.required }}</p>
      </div>
      <div class="form-group" :class="{error: $v.user.email.$error}">
        <label for="email">Email address</label>
        <input
          type="email"
          class="form-control"
          id="email"
          aria-describedby="email"
          placeholder="Enter your email"
          v-model.trim="$v.user.email.$model"
        >
        <p v-if="!$v.user.email.email">Please provide a valid email address</p>
        <p v-if="$v.user.name.$error && !$v.user.email.required">{{ errorMessages.required }}</p>
      </div>
      <div class="form-group" :class="{error: $v.user.phone.$error}">
        <label for="phone">Phone</label>
        <input
          type="text"
          class="form-control"
          id="phone"
          aria-describedby="phone"
          placeholder="Enter your phone number"
          v-model.trim="$v.user.phone.$model"
        >
        <p v-if="$v.user.name.$error && !$v.user.phone.required">{{ errorMessages.required }}</p>
      </div>
      Your gender:
      <div class="form-check form-check-inline">
        <input
          class="form-check-input"
          type="radio"
          name="gender"
          id="gender1"
          value="male"
          v-model="user.gender"
        >
        <label class="form-check-label" for="gender1">male</label>
      </div>
      <div class="form-check form-check-inline">
        <input
          class="form-check-input"
          type="radio"
          name="gender"
          id="gender2"
          value="female"
          v-model="user.gender"
        >
        <label class="form-check-label" for="gender2">female</label>
      </div>
      <div class="form-group" :class="{error: $v.user.password.$error}" v-if="formType === 'addNewUser'">
        <label for="password">Password</label>
        <input
          type="password"
          class="form-control"
          id="password"
          aria-describedby="password"
          placeholder="Choose a password (min 6 characters)"
          v-model.trim="$v.user.password.$model"
        >
        <p v-if="$v.user.password.$error && !$v.user.password.required">{{ errorMessages.required }}</p>
        <p v-if="!$v.user.password.minLength">Password must have at least {{ $v.user.password.$params.minLength.min }} letters.</p>
      </div>
      <div class="form-group" :class="{error: $v.user.passwordAgain.$error}" v-if="formType === 'addNewUser'">
        <label for="password-again">Repeat password</label>
        <input
          type="password"
          class="form-control"
          id="password-again"
          aria-describedby="password-again"
          placeholder="Enter your password again"
          v-model.trim="$v.user.passwordAgain.$model"
        >
        <p v-if="$v.user.passwordAgain.$error && !$v.user.passwordAgain.sameAsPassword">Passwords must be identical.</p>
      </div>
      Choose services:
      <div class="form-check form-check-inline">
        <input
          class="form-check-input"
          type="checkbox"
          value="mobile"
          id="service-mobile"
          v-model="user.services"
        >
        <label class="form-check-label" for="service-mobile">
          Mobile
        </label>
      </div>
      <div class="form-check form-check-inline">
        <input
          class="form-check-input"
          type="checkbox"
          value="internet"
          id="service-internet"
          v-model="user.services"
        >
        <label class="form-check-label" for="service-internet">
          Internet
        </label>
      </div>
      <div class="form-check form-check-inline">
        <input
          class="form-check-input"
          type="checkbox"
          value="tv"
          id="service-tv"
          v-model="user.services"
        >
        <label class="form-check-label" for="service-tv">
          TV
        </label>
      </div>
      <div class="form-group" :class="{error: !$v.user.country.required}">
        <label for="country">Select your country</label>
        <select
          class="form-control"
          id="country"
          v-model="$v.user.country.$model"
        >
          <option
            v-for="(country, index) in countries"
            :key="index"
            :value="country"
          >
          {{ country }}
          </option>
        </select>
        <p class="error" v-if="!$v.user.country.required">If you pick services country is required</p>
      </div>
      <div class="form-group">
        <div class="form-check">
          <input
            class="form-check-input"
            type="checkbox"
            id="newsletter"
            v-model="user.newsletter"
          >
          <label class="form-check-label" for="newsletter">
            Newsletter?
          </label>
        </div>
      </div>
      <div class="text-danger font-weight-bold py-3" v-if="error">{{ error }}</div>
      <button type="submit" class="btn btn-primary">Save</button>
    </form>
  </div>
</template>

<script>
import { email, required, sameAs, minLength, requiredIf } from 'vuelidate/lib/validators';
import formMixin from '../mixins/form';

export default {
  props: {
    userProp: {
      default: () => ({
        id: null,
        name: '',
        email: '',
        phone: '',
        gender: '',
        services: [],
        country: '',
        newsletter: false,
      }),
      type: Object,
    },
    formType: String,
  },
  mixins: [formMixin],
  data() {
    return {
      countries: ['Germany', 'Hungary', 'Italy', 'Spain', 'UK', 'USA'],
      user: {
        name: this.userProp.name,
        email: this.userProp.email,
        phone: this.userProp.phone,
        gender: this.userProp.gender,
        password: '',
        passwordAgain: '',
        services: this.userProp.services,
        country: this.userProp.country,
        newsletter: this.userProp.newsletter,
      },
      errorMessages: {
        required: 'Required field'
      }
    }
  },
  validations: {
    user: {
      name: {
        required,
      },
      email: {
        required,
        email,
      },
      phone: {
        required,
      },
      password: {
        required: requiredIf(function() {
          return this.formType === 'addNewUser';
        }),
        minLength: minLength(6),
      },
      passwordAgain: {
        sameAsPassword: sameAs('password'),
      },
      country: {
        required: requiredIf(vm => {
          return vm.services.length > 0;
        })
      }
    }
  },
  methods: {
    handleSaveUser() {
      this.$v.$touch();

      if (!this.$v.$invalid) {
        this.$emit('userData', {
          ...this.user,
          services: this.user.services.join(', '),
          id: this.userProp.id
        });
      }
    }
  },
  computed: {
    error() {
      return this.$store.getters.error;
    }
  }
}
</script>

<style lang="scss" scoped>
.error {
  color: red;
  input {
    border: 1px solid red;
  }
}
</style>
