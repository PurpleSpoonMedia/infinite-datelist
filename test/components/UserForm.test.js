import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import UserForm from '../../src/components/UserForm.vue';
import users from '../fixtures/users';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('UserForm.vue', () => {

  test('should render correctly when form is child of AddNewUser page', () => {
    let store, getters;
    getters = {
      error: () => false,
    };
    store = new Vuex.Store({
      getters
    });

    const wrapper = shallowMount(UserForm, {
      propsData: {
        formType: 'addNewUser'
      },
      store,
      localVue
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('#error-message').exists()).toBe(false);
  })

  test('should render correctly when form is child of EditUserData page', () => {
    let store, getters;
    getters = {
      error: () => false,
    };
    store = new Vuex.Store({
      getters
    });
    const wrapper = shallowMount(UserForm, {
      propsData: {
        formType: 'editUser',
        userProp: users[1]
      },
      store,
      localVue
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('#error-message').exists()).toBe(false);
  })

  test('should display error message', () => {
    let store, getters;
    getters = {
      error: () => 'Error message',
    };
    store = new Vuex.Store({
      getters
    });
    const wrapper = shallowMount(UserForm, {
      propsData: {
        formType: 'editUser',
        userProp: users[1]
      },
      store,
      localVue
    });
    expect(wrapper.find('#error-message').exists()).toBe(true);
  })

  test('should submit form without errors', () => {
    let store, getters;
    getters = {
      error: () => false,
    };
    store = new Vuex.Store({
      getters
    });

    const handleSaveUser = jest.fn();

    const wrapper = shallowMount(UserForm, {
      propsData: {
        formType: 'addNewUser'
      },
      methods: {
        handleSaveUser
      },
      store,
      localVue,
    });

    wrapper.setData({
      user: {
        ...users[0],
        password: '123456',
        passwordAgain: '123456'
      }
    });

    const form = wrapper.find('form');

    form.trigger('submit');
    expect(handleSaveUser).toHaveBeenCalled();
    expect(wrapper.findAll('.error').length).toBe(0);
  })
});
