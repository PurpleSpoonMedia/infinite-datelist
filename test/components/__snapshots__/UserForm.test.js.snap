// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm.vue should render correctly when form is child of AddNewUser page 1`] = `
<div class="px-lg-7">
  <form>
    <div class="form-group">
      <label for="name">Name</label>
      <input type="text" id="name" aria-describedby="name" placeholder="Enter your name" class="form-control">
      <!---->
    </div>
    <div class="form-group">
      <label for="email">Email address</label>
      <input type="email" id="email" aria-describedby="email" placeholder="Enter your email" class="form-control">
      <!---->
      <!---->
    </div>
    <div class="form-group">
      <label for="phone">Phone</label>
      <input type="text" id="phone" aria-describedby="phone" placeholder="Enter your phone number" class="form-control">
      <!---->
    </div>
    Your gender:
    <div class="form-check form-check-inline">
      <input type="radio" name="gender" id="gender1" value="male" class="form-check-input">
      <label for="gender1" class="form-check-label">male</label>
    </div>
    <div class="form-check form-check-inline">
      <input type="radio" name="gender" id="gender2" value="female" class="form-check-input">
      <label for="gender2" class="form-check-label">female</label>
    </div>
    <div class="form-group">
      <label for="password">Password</label>
      <input type="password" id="password" aria-describedby="password" placeholder="Choose a password (min 6 characters)" class="form-control">
      <!---->
      <!---->
    </div>
    <div class="form-group">
      <label for="password-again">Repeat password</label>
      <input type="password" id="password-again" aria-describedby="password-again" placeholder="Enter your password again" class="form-control">
      <!---->
    </div>
    Choose services:
    <div class="form-check form-check-inline">
      <input type="checkbox" value="mobile" id="service-mobile" class="form-check-input">
      <label for="service-mobile" class="form-check-label">
        Mobile
      </label>
    </div>
    <div class="form-check form-check-inline">
      <input type="checkbox" value="internet" id="service-internet" class="form-check-input">
      <label for="service-internet" class="form-check-label">
        Internet
      </label>
    </div>
    <div class="form-check form-check-inline">
      <input type="checkbox" value="tv" id="service-tv" class="form-check-input">
      <label for="service-tv" class="form-check-label">
        TV
      </label>
    </div>
    <div class="form-group">
      <label for="country">Select your country</label>
      <select id="country" class="form-control">
        <option value="Germany">
          Germany
        </option>
        <option value="Hungary">
          Hungary
        </option>
        <option value="Italy">
          Italy
        </option>
        <option value="Spain">
          Spain
        </option>
        <option value="UK">
          UK
        </option>
        <option value="USA">
          USA
        </option>
      </select>
      <!---->
    </div>
    <div class="form-group">
      <div class="form-check">
        <input type="checkbox" id="newsletter" class="form-check-input">
        <label for="newsletter" class="form-check-label">
          Newsletter?
        </label>
      </div>
    </div>
    <!---->
    <button id="form-submit" type="submit" class="btn btn-primary">Save</button>
  </form>
</div>
`;

exports[`UserForm.vue should render correctly when form is child of EditUserData page 1`] = `
<div class="px-lg-7">
  <form>
    <div class="form-group">
      <label for="name">Name</label>
      <input type="text" id="name" aria-describedby="name" placeholder="Enter your name" class="form-control">
      <!---->
    </div>
    <div class="form-group">
      <label for="email">Email address</label>
      <input type="email" id="email" aria-describedby="email" placeholder="Enter your email" class="form-control">
      <!---->
      <!---->
    </div>
    <div class="form-group">
      <label for="phone">Phone</label>
      <input type="text" id="phone" aria-describedby="phone" placeholder="Enter your phone number" class="form-control">
      <!---->
    </div>
    Your gender:
    <div class="form-check form-check-inline">
      <input type="radio" name="gender" id="gender1" value="male" class="form-check-input">
      <label for="gender1" class="form-check-label">male</label>
    </div>
    <div class="form-check form-check-inline">
      <input type="radio" name="gender" id="gender2" value="female" class="form-check-input">
      <label for="gender2" class="form-check-label">female</label>
    </div>
    <!---->
    <!---->
    Choose services:
    <div class="form-check form-check-inline">
      <input type="checkbox" value="mobile" id="service-mobile" class="form-check-input">
      <label for="service-mobile" class="form-check-label">
        Mobile
      </label>
    </div>
    <div class="form-check form-check-inline">
      <input type="checkbox" value="internet" id="service-internet" class="form-check-input">
      <label for="service-internet" class="form-check-label">
        Internet
      </label>
    </div>
    <div class="form-check form-check-inline">
      <input type="checkbox" value="tv" id="service-tv" class="form-check-input">
      <label for="service-tv" class="form-check-label">
        TV
      </label>
    </div>
    <div class="form-group">
      <label for="country">Select your country</label>
      <select id="country" class="form-control">
        <option value="Germany">
          Germany
        </option>
        <option value="Hungary">
          Hungary
        </option>
        <option value="Italy">
          Italy
        </option>
        <option value="Spain">
          Spain
        </option>
        <option value="UK">
          UK
        </option>
        <option value="USA">
          USA
        </option>
      </select>
      <!---->
    </div>
    <div class="form-group">
      <div class="form-check">
        <input type="checkbox" id="newsletter" class="form-check-input">
        <label for="newsletter" class="form-check-label">
          Newsletter?
        </label>
      </div>
    </div>
    <!---->
    <button id="form-submit" type="submit" class="btn btn-primary">Save</button>
  </form>
</div>
`;
